### DISCUSSION TITLE
about kafka message retention

### DISCUSSION THREAD

**Discussion by wookiist** [milvus-io/milvus#31765] [Q&A and General discussion]
How much of the kafka messages that milvus generates should we keep? 
I'm trying to set the value of `retention.ms` and was wondering if there is a recommended value. 
Thanks!

**Comment by wookiist** [comment]
Also, even if you set partition related settings in kafka topic, it seems that you only use 1 partition. Is this to solve the difficulty of sequential processing due to partitioning? It seems that there is no separate message key created by milvus.

**Comment by yhmo** [comment]
Kafka is like a WAL(write-ahead-log) component for Milvus. All the DML requests(insert/delete/upsert) are stored in Kafka before being processed by data nodes.

When the client sends an insert request, the proxy node receives the request, and passes the request to Kafka. Once the request is successfully stored into Kafka, the proxy will return a success status to the client. The request is not executed instantly, but it is guaranteed to be executed later.

If you have intensive/continually insert requests, the data nodes cannot keep up with the speed of kafka consuming the requests, Then there will be lots of requests accumulated in kafka. After you stop sending insert requests, data nodes will take a long time to process the requests. So, a long retention time of kafka will ensure that no requests are dropped without being processed. In this use case, I suggest to set the retention time value to 1 day.
If you have no intensive/continually insert requests, then you can choose a small retention time value. I think you can set it to 1 hour.

Yes, we only use one partition in kafka.

**Reply by wookiist** [reply]
@yhmo @xiaofan-luan Thanks to both of you as always. 🙇🏼 
I have an additional question. 
I don't see any consumer group subscribed to the milvus kafka topics, and I was wondering how the milvus component reads the kafka messages.

**Reply by xiaofan-luan** [reply]
both querynodes and datanodes read this message.
We predefined 256 topics, and if you have multiple shards, then multiple topics will be used.
If you don't see any consumers yet, which maybe means this shard is not used yet

**Reply by wookiist** [reply]
@xiaofan-luan  Are the 256 predefined topics you mentioned defined in kafka? I'm not sure I understand that part. 
I was wondering about the lack of consumer groups for these topics with the pattern of `by-dev-rootcoord-dml_X` in the default settings. 
Since this is just a WAL log, is it correct that there is no consumer?

**Reply by xiaofan-luan** [reply]
Nope, if the topic is actaully using, there should be one querynode and one datanode as consumer

**Reply by wookiist** [reply]
thank you @xiaofan-luan 
But there's still something I don't understand.
1. are the 256 predefined topics you mentioned auto-created in kafka? I have auto_creation set to False, so milvus can't create them by itself, only topics corresponding to the channel. 
2. and my question was that there is no consumer currently reading the message that came into the topic corresponding to the channel mentioned above.

**Reply by yhmo** [reply]
The topics are lazy-created by Milvus, using Kafka Go sdk. Create a collection then you will see the topic. Insert some data then you will see the query node and data node consume the messages.

**Reply by wookiist** [reply]
@yhmo @xiaofan-luan 
Thank you, but it's been 2 weeks since I started testing and the data is piling up and I can query it, 
still no consumer is identified. Is there any way to check the name of the consumer? As far as I can tell, [this function ](https://github.com/milvus-io/milvus/blob/master/pkg/mq/msgdispatcher/manager.go#L82-L84)seems to be creating the name of the consumer.

![image](https://github.com/milvus-io/milvus/assets/16011260/623c8478-2d8e-4b76-99cc-531b345a608a)
The image above is a milvus kafka topic that has accumulated a lot of data. 

<img width="1781" alt="image" src="https://github.com/milvus-io/milvus/assets/16011260/a31a5556-68d4-4d50-b1b2-2658f70ce814">
The image above shows a situation where no consumer still exists. 

I checked to see if the consumer offset is being recorded in etcd and managing its own offset, but that doesn't seem to be the case either. I'm wondering if there is a way to check to what extent the consumer has consumed the kafka topic. And could you tell me why it is not registered in the consumer group? Thank you very much. 🙇🏼

**Reply by wookiist** [reply]
https://github.com/milvus-io/milvus/blob/ae8f46f431379ab921445e2912081d907dcf5742/internal/proto/etcd_meta.proto#L52-L71

**Reply by xiaofan-luan** [reply]
Producer is rootcoord and proxy.
Consumer is the datanode and querynode consume this topic. It is assigned by querycoord and datacoord.

**Reply by xiaofan-luan** [reply]
I'm not 100% understand, why are you care about the producer and consumer ?
I think you will need to config your kafka config for the retention time instead

**Reply by wookiist** [reply]
We wanted to check the consumer group to see how much lag there was between the data accumulated in MQ and the data read by the datanode on insert.
After our own investigation, we found that milvus does not use a separate consumer group, but rather has its own offset management. 
However, I still haven't figured out how to check the lag, can you help me with that? 
Thanks for your kind help as always.

**Reply by xiaofan-luan** [reply]
in milvus, there is something called timetick lag, it is also showed in grafana dash board.
It tells you how much lag you have on consumer side

**Comment by xiaofan-luan** [comment]
each shard use 1 topic, each topic has only one partition. 
And yes, this is related to the message ordering.
You can increasing shard number to increase write throughput

**Comment by jaime0815** [comment]
the 1 topic will be assigned when creating the collection with 1 shard that topic might be new or exists. 
you can use kafka-topics.sh to check topic lists,  topic by-dev-rootcoord-dml_0  and by-dev-rootcoord-delta_0 are created after creating 1 shard collection.
```
bin/kafka-topics.sh --bootstrap-server localhost:9092 --list
__consumer_offsets
by-dev-rootcoord-delta_0
by-dev-rootcoord-dml_0
```
The default retention configuration is one week that can be configured with global level or topic level,  using kafka-configs.sh to check topic retention configuration:
```
 bin/kafka-configs.sh  --bootstrap-server localhost:9092  --topic by-dev-rootcoord-dml_0 --describe  --all | grep retention
  retention.ms=604800000 sensitive=false synonyms={}
  retention.bytes=-1 sensitive=false synonyms={DEFAULT_CONFIG:log.retention.bytes=-1}
  delete.retention.ms=86400000 sensitive=false synonyms={DEFAULT_CONFIG:log.cleaner.delete.retention.ms=86400000}
```

using default retention configuration is sufficient,  and due to Milvus not supporting set retention on topic level that can be configured with global level in kafka cluster.
